java/vector/src/main/codegen/templates/DenseUnionVector.java

```diff
@@ -0,0 +1,887 @@
+
+  @Override
+  public int getBufferSizeFor(final int valueCount) {
+    if (valueCount == 0) {
+      return 0;
+    }
+
+    long bufferSize = 0;
+    for (final ValueVector v : (Iterable<ValueVector>) this) {
+      bufferSize += v.getBufferSizeFor(valueCount);
+    }
+
+    return (int) bufferSize + valueCount * TYPE_WIDTH + valueCount * OFFSET_WIDTH;
+  }
+
+  @Override
+  public ArrowBuf[] getBuffers(boolean clear) {
+    List<ArrowBuf> list = new java.util.ArrayList<>();
+    setReaderAndWriterIndex();
+    if (getBufferSize() != 0) {
+      list.add(typeBuffer);
+      list.add(offsetBuffer);
+      list.addAll(java.util.Arrays.asList(internalStruct.getBuffers(clear)));
+    }
+    if (clear) {
+      valueCount = 0;
+      typeBuffer.getReferenceManager().retain();
+      typeBuffer.close();
+      typeBuffer = allocator.getEmpty();
+      offsetBuffer.getReferenceManager().retain();
+      offsetBuffer.close();
+      offsetBuffer = allocator.getEmpty();
+    }
+    return list.toArray(new ArrowBuf[list.size()]);
+  }
+
+  @Override
+  public Iterator<ValueVector> iterator() {
+    List<ValueVector> vectors = org.apache.arrow.util.Collections2.toList(internalStruct.iterator());
+    return vectors.iterator();
+  }
+
+  private ValueVector getVector(int index) {
+    int type = typeBuffer.getByte(index * TYPE_WIDTH);
+    if (type < Types.MinorType.values().length) {
+      switch (MinorType.values()[type]) {
+        case NULL:
+          return null;
+      <#list vv.types as type>
+        <#list type.minor as minor>
+          <#assign name = minor.class?cap_first />
+          <#assign fields = minor.fields !type.fields />
+          <#assign uncappedName = name?uncap_first />
+          <#if !minor.typeParams ??>
+        case ${name?upper_case}:
+        return get${name}Vector();
+          </#if>
+        </#list>
+      </#list>
+        default:
+          throw new UnsupportedOperationException("Cannot support type: " + MinorType.values()[type]);
+      }
+    } else {
+      if (typeFields[type].getType() instanceof Struct) {
+        return getStruct(type);
+      } else if (typeFields[type].getType() instanceof List) {
+        return getList(type);
+      } else {
+        throw new UnsupportedOperationException("Cannot support type: " + type);
+      }
+    }
+  }
+
+  public Object getObject(int index) {
+    ValueVector vector = getVector(index);
+    if (vector != null) {
+      int offset = offsetBuffer.getInt(index * OFFSET_WIDTH);
+      return vector.getObject(offset);
+    }
+    return null;
+  }
+
+  public void get(int index, DenseUnionHolder holder) {
+    FieldReader reader = new DenseUnionReader(DenseUnionVector.this);
+    int offset = offsetBuffer.getInt(index * OFFSET_WIDTH);
+    reader.setPosition(offset);
+    holder.reader = reader;
+  }
+
+  public int getValueCount() {
+    return valueCount;
+  }
+
+  public boolean isNull(int index) {
+    return (typeBuffer.getByte(index * TYPE_WIDTH) == 0);
```

Comment:
there should be a separate validity buffer or there needs to be a more discussion the mailing list.  As the spec is currently written this doesn't conform.

Reply:
This is a real problem, and UnionVector also had this problem. 
We solved it by using the underlying vector's validity buffer (please see https://github.com/apache/arrow/pull/5752#discussion_r340437254).

Reply:
After more thoughts, I think the solution for UnionVector does not work for DenseUnionVector. So I have revised the code to add a validity buffer.